Kafka kraft

sschubertchainguard opened the pull request:
Fixes:
- Relocate server.properties to address issue when deploying with KRaft mode
- Add docker-entrypoint-initdb.d dir to quiet warning

sschubertchainguard committed:
fix kafka kraft mode issue

Signed-off-by: Scott Schubert <scott.schubert@chainguard.dev>

sschubertchainguard committed:
fix kafka kraft mode issue

Signed-off-by: Scott Schubert <scott.schubert@chainguard.dev>

kranurag7 commented on kafka-3.8.yaml:
```diff
@@ -73,6 +73,8 @@ subpackages:
           ln -s /usr/lib/kafka/config ${{targets.subpkgdir}}/opt/bitnami/kafka/config
           ln -s /usr/lib/kafka/logs ${{targets.subpkgdir}}/opt/bitnami/kafka/logs
+          mv ${{targets.destdir}}/usr/lib/kafka/config/server.properties ${{targets.destdir}}/usr/lib/kafka/config/server.properties.original
+          mkdir ${{targets.destdir}}/docker-entrypoint-initdb.d/
```
the directory seems specific to `bitnami-compat` I think we should use `${{targets.contextdir}}` here.

sschubertchainguard committed:
update to subpkgdir

Signed-off-by: Scott Schubert <scott.schubert@chainguard.dev>

kranurag7 commented on kafka-3.8.yaml:
```diff
@@ -73,6 +73,8 @@ subpackages:
           ln -s /usr/lib/kafka/libs ${{targets.subpkgdir}}/opt/bitnami/kafka/libs
           ln -s /usr/lib/kafka/config ${{targets.subpkgdir}}/opt/bitnami/kafka/config
           ln -s /usr/lib/kafka/logs ${{targets.subpkgdir}}/opt/bitnami/kafka/logs
+          mv ${{targets.contextdir}}/usr/lib/kafka/config/server.properties ${{targets.contextdir}}/usr/lib/kafka/config/server.properties.original
```
a quick fs lookup on `bitnami/kafka:3.8` image shows me that the file is located at some other place.

```bash
 $ fd properties.original
opt/bitnami/kafka/config/server.properties.original
```
Is there a reason to put it inside `/usr/lib` here? 

I think we should proceed with the following: 

```bash
cp /usr/lib/kafka/config/server.properties ${{targets.contextdir}}/opt/bitnami/kafka/config/server.properties.original
```

developer-guy replied:
I'm also not certain that contextdir works in that situation 🤔 so it could be targets.destdir to targets.subpkgdir instead of contextdir

kranurag7 replied:
I think given it's under `subpackages` I think contextdir should figure out that given it's a subpackage it should be `subpkgdir` 

Looking at other bitnami/compat usage, I think we are good with contextdir as well but we can move to subpkgdir for sure to be more specific. 

Also, I think given you're copying this file `usr/lib/kafka/config/server.properties` from main package pipeline to subpackage one, you can either use `cp /usr/lib/kafka/config/server.properties ${{targets.contextdir}}/opt/bitnami/kafka/config/server.properties.original` or `cp ${{targets.destdir}}/usr/lib/kafka/config/server.properties ${{targets.contextdir}}/opt/bitnami/kafka/config/server.properties.original` 

@developer-guy please let me know if I missed something here.

sschubertchainguard replied:
Locally I've tried:
```
# encounters ...WARN mv: can't rename '/home/build/melange-out/kafka-3.8/usr/lib/kafka/config/server.properties': No such file or directory
mv ${{targets.destdir}}/usr/lib/kafka/config/server.properties ${{targets.contextdir}}/usr/lib/kafka/config/server.properties.original
mv ${{targets.destdir}}/usr/lib/kafka/config/server.properties ${{targets.subpkgdir}}/usr/lib/kafka/config/server.properties.original

# encounters ...WARN mv: can't rename '/usr/lib/kafka/config/server.properties': No such file or directory
mv /usr/lib/kafka/config/server.properties ${{targets.subpkgdir}}/usr/lib/kafka/config/server.properties.original
```

So far can only get the following to build successfully:
```
mv ${{targets.destdir}}/usr/lib/kafka/config/server.properties ${{targets.destdir}}/usr/lib/kafka/config/server.properties.original
```

sschubertchainguard committed:
update to opt/bitnami

Signed-off-by: Scott Schubert <scott.schubert@chainguard.dev>

sschubertchainguard committed:
fix paths and vars

Signed-off-by: Scott Schubert <scott.schubert@chainguard.dev>

kranurag7 committed:
rename to server.properties.original in bitnami-compat

the application in kraft mode requires this.

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
copy things into bitnami package

a direct symlink breaks the application because of presence of a file,
the way to handle this as of now is to copy things and rename the file
that is causing error.
this commit does the same.

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

kranurag7 committed:
add runtime deps to bitnami-compat

this is required  because now for the bitnami variant of this image we
will import this package.

Signed-off-by: kranurag7 <81210977+kranurag7@users.noreply.github.com>

sschubertchainguard committed:
add kafka-bitnami-kraft-compat package

Signed-off-by: Scott Schubert <scott.schubert@chainguard.dev>

sschubertchainguard committed:
Merge branch 'wolfi-dev:main' into kafka-kraft

sschubertchainguard committed:
add kafka-bitnami-kraft-compat package

Signed-off-by: Scott Schubert <scott.schubert@chainguard.dev>

kranurag7 commented on kafka-3.8.yaml:
```diff
@@ -70,6 +70,30 @@ subpackages:
           version-path: 3.8/debian-12
       - runs: |
+<<<<<<< HEAD
+=======
```
this and one more line below can be removed. Sorry if you're already on it.

sschubertchainguard replied:
Resolved - sorry for the noise!

sschubertchainguard committed:
add kafka-bitnami-kraft-compat package

Signed-off-by: Scott Schubert <scott.schubert@chainguard.dev>

xnox commented on kafka-3.8.yaml:
```diff
@@ -74,6 +74,31 @@ subpackages:
+          cp -r ${{targets.destdir}}/usr/lib/kafka/libs ${{targets.subpkgdir}}/opt/bitnami/kafka
+          cp -r ${{targets.destdir}}/usr/lib/kafka/logs ${{targets.subpkgdir}}/opt/bitnami/kafka
+          cp -r ${{targets.destdir}}/usr/lib/kafka/config ${{targets.subpkgdir}}/opt/bitnami/kafka
+          mv ${{targets.subpkgdir}}/opt/bitnami/kafka/config/server.properties ${{targets.subpkgdir}}/opt/bitnami/kafka/config/server.properties.original
```
in bitnami image i think all modes are co-located, do we really need a separate build of compat package; then then also a separate build of image?

As in, can this all not be folded into kafka-bitnami-compat-${{vars.major-minor-version}} which we now know is incomplete?

kranurag7 replied:
Thanks for the review. I think this PR is superseded via https://github.com/wolfi-dev/os/pull/29341 (thanks for @srishtih) 

It was an ownership issue so far we tracked it. 

We are renaming the file `server.properties` to `server.properties.original` in the main package and when we are running it then `server.properties` file is getting created at runtime with `kafka:kafka` ownership while the application runs.

After renaming things, docker-compose setup, package level and image level tests are passing so we can drop this extra `compat` package which we were building for kafka kraft mode.

xnox replied:
i like that other PR better it looks smaller! but it is marked as draft. Reviewed that other one, and will park this one for now.

sschubertchainguard commented:
Closing in favor of https://github.com/wolfi-dev/os/pull/29341
